src/hotspot/cpu/x86/stubGenerator_x86_64_poly.cpp

```diff
@@ -0,0 +1,1027 @@
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "macroAssembler_x86.hpp"
+#include "stubGenerator_x86_64.hpp"
+
+#define __ _masm->
+
+// References:
+//  - (Normative) RFC7539 - ChaCha20 and Poly1305 for IETF Protocols
+//  - M. Goll and S. Gueron, "Vectorization of Poly1305 Message Authentication Code"
+//  - "The design of Poly1305" https://loup-vaillant.fr/tutorials/poly1305-design
+
+// Explanation for the 'well known' modular arithmetic optimization, reduction by pseudo-Mersene prime 2^130-5:
+//
+// Reduction by 2^130-5 can be expressed as follows:
+//    ( a×2^130 + b ) mod 2^130-5     //i.e. number split along the 130-bit boundary
+//                                 = ( a×2^130 - 5×a + 5×a + b ) mod 2^130-5
+//                                 = ( a×(2^130 - 5) + 5×a + b ) mod 2^130-5 // i.e. adding multiples of modulus is a noop
+//                                 = ( 5×a + b ) mod 2^130-5
+// QED: shows mathematically the well known algorithm of 'split the number down the middle, multiply upper and add'
+// This is particularly useful to understand when combining with 'odd-sized' limbs that might cause misallignment
+//
+
+// Pseudocode for this file (in general):
+//    * used for poly1305_multiply_scalar
+//    × used for poly1305_multiply8_avx512
+//    lower-case variables are scalar numbers in 3×44-bit limbs (in gprs)
+//    upper-case variables are 8-element vector numbers in 3×44-bit limbs (in zmm registers)
+//    [ ] used to denote vector numbers (with their elements)
+
+// Register Map:
+// GPRs:
+//   input        = rdi
+//   length       = rbx
+//   accumulator  = rcx
+//   R   = r8
+//   a0  = rsi
+//   a1  = r9
+//   a2  = r10
+//   r0  = r11
+//   r1  = r12
+//   c1  = r8;
+//   t1  = r13
+//   t2  = r14
+//   t3  = r15
+//   t0  = r14
+//   polyCP = r13
+//   stack(rsp, rbp)
+//   imul(rax, rdx)
+// ZMMs:
+//   T: xmm0-6
+//   C: xmm7-9
+//   A: xmm13-18
+//   B: xmm19-24
+//   R: xmm25-29
+
+// Constant Pool Offsets:
+enum polyCPOffset {
+  high_bit = 0,
+  mask_44 = 64,
+  mask_42 = 128,
+};
+ATTRIBUTE_ALIGNED(64) uint64_t POLY1305_CP[] = {
+  // OFFSET 0: high_bit
+  0x0000010000000000, 0x0000010000000000,
+  0x0000010000000000, 0x0000010000000000,
+  0x0000010000000000, 0x0000010000000000,
+  0x0000010000000000, 0x0000010000000000,
+
+  // OFFSET 64: mask_44
+  0xfffffffffff, 0xfffffffffff,
```

Comment:
Please, keep leading zeroes explicit in the constants.

Reply:
done. Also split things up and added ExternalAddress version of instructions.